diff --git a/kong.conf.default b/kong.conf.default
index 6ea2eb872..ed42400c7 100644
--- a/kong.conf.default
+++ b/kong.conf.default
@@ -148,7 +148,7 @@
                                  # instead of reporting the port Kong is
                                  # listening to.
 
-#anonymous_reports = on          # Send anonymous usage data such as error
+#anonymous_reports = off         # Send anonymous usage data such as error
                                  # stack traces to help improve Kong.
 
 #------------------------------------------------------------------------------
diff --git a/kong/templates/kong_defaults.lua b/kong/templates/kong_defaults.lua
index 1ee4899a7..afce6d506 100644
--- a/kong/templates/kong_defaults.lua
+++ b/kong/templates/kong_defaults.lua
@@ -12,7 +12,7 @@ go_pluginserver_exe = /usr/local/bin/go-pluginserver
 go_plugins_dir = off
 port_maps = NONE
 host_ports = NONE
-anonymous_reports = on
+anonymous_reports = off
 enforce_rbac = off
 rbac_auth_header = Kong-Admin-Token
 vitals = on
