diff --git a/LICENSE b/LICENSE
index 641bbad70..350b582bf 100644
--- a/LICENSE
+++ b/LICENSE
@@ -1 +1,3 @@
 Use of this software is subject to the terms of your license agreement with Kong Inc.
+
+Use, reproduction, or disclosure is subject to restrictions set forth in Contract Number 2014-14031000011 and Sub Contract No. Kong 01.
diff --git a/kong/api/routes/kong.lua b/kong/api/routes/kong.lua
index 839d77eba..c6d09bd46 100644
--- a/kong/api/routes/kong.lua
+++ b/kong/api/routes/kong.lua
@@ -21,7 +21,7 @@ local ERR = ngx.ERR
 local errors = Errors.new()
 
 
-local tagline = "Welcome to " .. _KONG._NAME
+local tagline = "Welcome to " .. _KONG._NAME .. ". Use, reproduction, or disclosure is subject to restrictions set forth in Contract Number 2014-14031000011 and Sub Contract No. Kong 01."
 local version = _KONG._VERSION
 local lua_version = jit and jit.version or _VERSION
 

