diff --git a/bundle/lua-resty-dns-0.22/lib/resty/dns/resolver.lua b/bundle/lua-resty-dns-0.22/lib/resty/dns/resolver.lua
index a67b3c1..0305485 100644
--- a/bundle/lua-resty-dns-0.22/lib/resty/dns/resolver.lua
+++ b/bundle/lua-resty-dns-0.22/lib/resty/dns/resolver.lua
@@ -99,6 +99,26 @@ for i = 2, 64, 2 do
     arpa_tmpl[i] = DOT_CHAR
 end

+local function udp_socks_close(self)
+    if self.socks == nil then
+        return
+    end
+
+    for _, sock in ipairs(self.socks) do
+        sock:close()
+    end
+
+    self.socks = nil
+end
+
+local function tcp_socks_close(self)
+    if self.tcp_sock == nil then
+        return
+    end
+
+    self.tcp_sock:close()
+    self.tcp_sock = nil
+end

 function _M.new(class, opts)
     if not opts then
@@ -161,6 +181,14 @@ function _M.new(class, opts)
                 }, mt)
 end

+function _M:destroy()
+    udp_socks_close(self)
+    tcp_socks_close(self)
+    self.cur = nil
+    self.servers = nil
+    self.retrans = nil
+    self.no_recurse = nil
+end

 local function pick_sock(self, socks)
     local cur = self.cur
