diff --git a/bundle/LuaJIT-2.1-20210510/src/luajit.h b/bundle/LuaJIT-2.1-20210510/src/luajit.h
index 1dddaec..651de67 100644
--- a/bundle/LuaJIT-2.1-20210510/src/luajit.h
+++ b/bundle/LuaJIT-2.1-20210510/src/luajit.h
@@ -32,7 +32,9 @@
 
 #define OPENRESTY_LUAJIT
 
+#ifndef LUAJIT_VERSION
 #define LUAJIT_VERSION		"LuaJIT 2.1.0-beta3"
+#endif
 #define LUAJIT_VERSION_NUM	20100  /* Version 2.1.0 = 02.01.00. */
 #define LUAJIT_VERSION_SYM	luaJIT_version_2_1_0_beta3
 #define LUAJIT_COPYRIGHT	"Copyright (C) 2005-2021 Mike Pall"
