diff --git a/bundle/LuaJIT-2.1-20231117.1/src/luajit_rolling.h b/bundle/LuaJIT-2.1-20231117.1/src/luajit_rolling.h
index f082974..d16d66b 100644
--- a/bundle/LuaJIT-2.1-20231117.1/src/luajit_rolling.h
+++ b/bundle/LuaJIT-2.1-20231117.1/src/luajit_rolling.h
@@ -32,7 +32,9 @@
 
 #define OPENRESTY_LUAJIT
 
+#ifndef LUAJIT_VERSION
 #define LUAJIT_VERSION		"LuaJIT 2.1.ROLLING"
+#endif
 #define LUAJIT_VERSION_NUM	20199  /* Deprecated. */
 #define LUAJIT_VERSION_SYM	luaJIT_version_2_1_ROLLING
 #define LUAJIT_COPYRIGHT	"Copyright (C) 2005-2023 Mike Pall"
