From f53c8fa441f4233b9a3f19fcd870207fe8795456 Mon Sep 17 00:00:00 2001
From: Qi <add_sp@outlook.com>
Date: Wed, 25 May 2022 18:35:08 +0800
Subject: [PATCH] Patch macro `LUAJIT_VERSION`

---
 src/luajit.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/bundle/LuaJIT-2.1-20220411/src/luajit.h b/bundle/LuaJIT-2.1-20220411/src/luajit.h
index a4d33001..e35f4e7e 100644
--- a/bundle/LuaJIT-2.1-20220411/src/luajit.h
+++ b/bundle/LuaJIT-2.1-20220411/src/luajit.h
@@ -32,7 +32,9 @@

 #define OPENRESTY_LUAJIT

+#ifndef LUAJIT_VERSION
 #define LUAJIT_VERSION		"LuaJIT 2.1.0-beta3"
+#endif
 #define LUAJIT_VERSION_NUM	20100  /* Version 2.1.0 = 02.01.00. */
 #define LUAJIT_VERSION_SYM	luaJIT_version_2_1_0_beta3
 #define LUAJIT_COPYRIGHT	"Copyright (C) 2005-2022 Mike Pall"
--
2.34.1


