diff --git a/bundle/nginx-1.25.3/src/http/modules/ngx_http_grpc_module.c b/bundle/nginx-1.25.3/src/http/modules/ngx_http_grpc_module.c
index dfe49c5..db85ca3 100644
--- a/bundle/nginx-1.25.3/src/http/modules/ngx_http_grpc_module.c
+++ b/bundle/nginx-1.25.3/src/http/modules/ngx_http_grpc_module.c
@@ -8,6 +8,9 @@
 #include <ngx_config.h>
 #include <ngx_core.h>
 #include <ngx_http.h>
+#if (NGX_HTTP_LUA_KONG)
+#include <ngx_http_lua_kong_module.h>
+#endif
 
 
 typedef struct {
@@ -733,6 +736,10 @@ ngx_http_grpc_create_request(ngx_http_request_t *r)
     len = sizeof(ngx_http_grpc_connection_start) - 1
           + sizeof(ngx_http_grpc_frame_t);             /* headers frame */
 
+#if (NGX_HTTP_LUA_KONG)
+    ngx_http_lua_kong_set_grpc_authority(r, &ctx->host);
+#endif
+
     /* :method header */
 
     if (r->method == NGX_HTTP_GET || r->method == NGX_HTTP_POST) {
