diff --git a/bundle/nginx-1.27.1/src/stream/ngx_stream_ssl_preread_module.c b/bundle/nginx-1.27.1/src/stream/ngx_stream_ssl_preread_module.c
index ed45c6f..7e65d65 100644
--- a/bundle/nginx-1.27.1/src/stream/ngx_stream_ssl_preread_module.c
+++ b/bundle/nginx-1.27.1/src/stream/ngx_stream_ssl_preread_module.c
@@ -190,7 +190,7 @@ ngx_stream_ssl_preread_handler(ngx_stream_session_t *s)
         }
 
         if (rc == NGX_OK) {
-            return ngx_stream_ssl_preread_servername(s, &ctx->host);
+            rc = ngx_stream_ssl_preread_servername(s, &ctx->host);
         }
 
         if (rc != NGX_AGAIN) {
