diff --git a/bundle/nginx-1.19.9/src/event/ngx_event_openssl.c b/bundle/nginx-1.19.9/src/event/ngx_event_openssl.c
index 6361810..54ff114 100644
--- a/bundle/nginx-1.19.9/src/event/ngx_event_openssl.c
+++ b/bundle/nginx-1.19.9/src/event/ngx_event_openssl.c
@@ -378,6 +378,10 @@ ngx_ssl_create(ngx_ssl_t *ssl, ngx_uint_t protocols, void *data)
     SSL_CTX_set_options(ssl->ctx, SSL_OP_NO_CLIENT_RENEGOTIATION);
 #endif
 
+#ifdef SSL_OP_IGNORE_UNEXPECTED_EOF
+    SSL_CTX_set_options(ssl->ctx, SSL_OP_IGNORE_UNEXPECTED_EOF);
+#endif
+
 #ifdef SSL_MODE_RELEASE_BUFFERS
     SSL_CTX_set_mode(ssl->ctx, SSL_MODE_RELEASE_BUFFERS);
 #endif
