diff --git a/bundle/nginx-1.25.3/src/http/ngx_http_core_module.c b/bundle/nginx-1.25.3/src/http/ngx_http_core_module.c
index 7845f8f..1676116 100644
--- a/bundle/nginx-1.25.3/src/http/ngx_http_core_module.c
+++ b/bundle/nginx-1.25.3/src/http/ngx_http_core_module.c
@@ -4184,7 +4184,7 @@ ngx_http_core_listen(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
 
         if (ngx_strcmp(value[n].data, "http2") == 0) {
 #if (NGX_HTTP_V2)
-            ngx_conf_log_error(NGX_LOG_WARN, cf, 0,
+            ngx_conf_log_error(NGX_LOG_NOTICE, cf, 0,
                                "the \"listen ... http2\" directive "
                                "is deprecated, use "
                                "the \"http2\" directive instead");
