diff --git a/bundle/ngx_stream_lua-0.0.16/src/api/ngx_stream_lua_api.h b/bundle/ngx_stream_lua-0.0.16/src/api/ngx_stream_lua_api.h
index 515cfd3..c743dae 100644
--- a/bundle/ngx_stream_lua-0.0.16/src/api/ngx_stream_lua_api.h
+++ b/bundle/ngx_stream_lua-0.0.16/src/api/ngx_stream_lua_api.h
@@ -21,6 +21,9 @@
 
 
 
+#include <ngx_stream.h>
+#include "../ngx_stream_lua_request.h"
+
 
 #include <lua.h>
 #include <stdint.h>
