diff --git a/bundle/ngx_lua-0.10.28/src/ngx_http_lua_log.c b/bundle/ngx_lua-0.10.28/src/ngx_http_lua_log.c
index 43ab8209..a83cc2de 100644
--- a/bundle/ngx_lua-0.10.28/src/ngx_http_lua_log.c
+++ b/bundle/ngx_lua-0.10.28/src/ngx_http_lua_log.c
@@ -253,10 +253,12 @@ log_wrapper(ngx_log_t *log, const char *ident, ngx_uint_t level,
                 break;

             case LUA_TLIGHTUSERDATA:
-                *p++ = 'n';
-                *p++ = 'u';
-                *p++ = 'l';
-                *p++ = 'l';
+                if (lua_touserdata(L, i) == NULL) {
+                    *p++ = 'n';
+                    *p++ = 'u';
+                    *p++ = 'l';
+                    *p++ = 'l';
+                }

                 break;

