From 9ce0848cff7c3c5eb0a7d5adfe2de22ea98e1e63 Mon Sep 17 00:00:00 2001
From: attenuation <ouyangjun1999@gmail.com>
Date: Sun, 21 Aug 2022 22:12:22 +0800
Subject: [PATCH] feat: add ngx_stream_lua_ffi_worker_pids to get all workers
 pid map

---

diff --git a/bundle/ngx_stream_lua-0.0.10/src/ngx_stream_lua_worker.c b/bundle/ngx_stream_lua-0.0.10/src/ngx_stream_lua_worker.c
index e4a9c915..4fb23314 100644
--- a/bundle/ngx_stream_lua-0.0.10/src/ngx_stream_lua_worker.c
+++ b/bundle/ngx_stream_lua-0.0.10/src/ngx_stream_lua_worker.c
@@ -28,6 +28,36 @@ ngx_stream_lua_ffi_worker_pid(void)
 }
 
 
+int
+ngx_stream_lua_ffi_worker_pids(int *pids, size_t *pids_len)
+{
+    ngx_int_t i, n;
+
+    n = 0;
+    for (i = 0; i < NGX_MAX_PROCESSES; i++) {
+        if (i != ngx_process_slot && ngx_processes[i].pid == 0) {
+            break;
+        }
+
+        if (i == ngx_process_slot && ngx_processes[i].pid == 0) {
+            pids[n++] = ngx_pid;
+        }
+
+        if (ngx_processes[i].pid > 0) {
+            pids[n++] = ngx_processes[i].pid;
+        }
+    }
+
+    if (n == 0) {
+        return NGX_ERROR;
+    }
+
+    *pids_len = n;
+
+    return NGX_OK;
+}
+
+
 int
 ngx_stream_lua_ffi_worker_id(void)
 {