diff --git a/bundle/ngx_lua-0.10.28/src/ngx_http_lua_ssl.h b/bundle/ngx_lua-0.10.28/src/ngx_http_lua_ssl.h
index 3d577c6..aa20f03 100644
--- a/bundle/ngx_lua-0.10.28/src/ngx_http_lua_ssl.h
+++ b/bundle/ngx_lua-0.10.28/src/ngx_http_lua_ssl.h
@@ -38,6 +38,9 @@ typedef struct {
     unsigned                 entered_client_hello_handler:1;
     unsigned                 entered_cert_handler:1;
     unsigned                 entered_sess_fetch_handler:1;
+#if (NGX_HTTP_LUA_KONG)
+    unsigned                 disable_http2_alpn:1;
+#endif
 } ngx_http_lua_ssl_ctx_t;
 
 
